### ISSUE TITLE
Documented Technique to Autocomplete Orders does not always work

### ISSUE THREAD

**Issue by JohnChandlerEdmonton** [woothemes/woocommerce#35629]
### Prerequisites

- [x] I have carried out troubleshooting steps and I believe I have found a bug.
- [X] I have searched for similar bugs in both open and closed issues and cannot find a duplicate.

### Describe the bug

The technique shown in the Woocommerce documentation on this page:
https://woocommerce.com/document/automatically-complete-orders/
does not always work.

*****RECOMMENDATION: ***The documentation should be changed slightly and then it will always work.  Instead of using

     add_action( 'woocommerce_thankyou', 'custom_woocommerce_auto_complete_order');

(which may not be executed because it possible to pay for an order but then stop before you go to the thank you page), I believe this is a better hook to use instead:

     add_action( 'woocommerce_checkout_order_processed', 'custom_woocommerce_auto_complete_order'); 

which will execute the code when it is needed no matter what.

P.S. Another problem with the thank-you page action is that if the user refreshes the page it will execute that action again.   'woocommerce_checkout_order_processed' does not have this problem. It will only be executed once, when the order is created.

### Expected behavior

When the code snippet shown in the documentation is put into functions.php then it should always set the order status to "completed" when a new order is created.

### Actual behavior

When the code snippet shown in the documentation is put into functions.php it may not set the order status to "completed" as expected.  This is because it is possible to pay for an order but then stop before you go to the thank you page.

### Steps to reproduce

1. Install Woocommerce
2. Create a product.
3. Activate the Paypal Standard gateway
4. Insert the code snippet found here  https://woocommerce.com/document/automatically-complete-orders/ into functions.php
5. Add the product to the cart
6. Checkout and pay with paypal.
7.  Do not click "Return to Merchant".  Instead just close the browser or go somewhere else.
8.  You will then see that the order's status is still "Processing".  It did not get set to "Completed."  This is because you never got to the "Thank You" page and so the code snippet was never executed.

### WordPress Environment

`
### WordPress Environment ###

WordPress address (URL): https://woocommerce-813784-2946681.cloudwaysapps.com
Site address (URL): https://woocommerce-813784-2946681.cloudwaysapps.com
WC Version: 7.1.0
REST API Version: ✔ 7.1.0
WC Blocks Version: ✔ 8.7.5
Action Scheduler Version: ✔ 3.4.0
Log Directory Writable: ✔
WP Version: 6.1.1
WP Multisite: –
WP Memory Limit: 256 MB
WP Debug Mode: –
WP Cron: ✔
Language: en_US
External object cache: –

### Server Environment ###

Server Info: Apache/2.4.54 (Debian)
PHP Version: 7.4.27
PHP Post Max Size: 301 MB
PHP Time Limit: 300
PHP Max Input Vars: 10000
cURL Version: 7.64.0
OpenSSL/1.1.1d

SUHOSIN Installed: –
MySQL Version: 5.5.5-10.4.20-MariaDB-1:10.4.20+maria~buster-log
Max Upload Size: 301 MB
Default Timezone is UTC: ✔
fsockopen/cURL: ✔
SoapClient: ✔
DOMDocument: ✔
GZip: ✔
Multibyte String: ✔
Remote Post: ✔
Remote Get: ✔

### Database ###

WC Database Version: 7.1.0
WC Database Prefix: wp_
Total Database Size: 6.30MB
Database Data Size: 4.52MB
Database Index Size: 1.78MB
wp_woocommerce_sessions: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_api_keys: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_attribute_taxonomies: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_downloadable_product_permissions: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_woocommerce_order_items: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_order_itemmeta: Data: 0.05MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_tax_rates: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_woocommerce_tax_rate_locations: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_shipping_zones: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_woocommerce_shipping_zone_locations: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_shipping_zone_methods: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_woocommerce_payment_tokens: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_payment_tokenmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_log: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_actions: Data: 0.06MB + Index: 0.11MB + Engine InnoDB
wp_actionscheduler_claims: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_groups: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_logs: Data: 0.05MB + Index: 0.03MB + Engine InnoDB
wp_bv_fw_requests: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_bv_ip_store: Data: 0.27MB + Index: 0.31MB + Engine InnoDB
wp_bv_lp_requests: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_commentmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_comments: Data: 0.02MB + Index: 0.09MB + Engine InnoDB
wp_links: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_options: Data: 3.08MB + Index: 0.06MB + Engine InnoDB
wp_ph_bookings_availability_calculation_data: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_postmeta: Data: 0.06MB + Index: 0.03MB + Engine InnoDB
wp_posts: Data: 0.05MB + Index: 0.06MB + Engine InnoDB
wp_termmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_terms: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_term_relationships: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_term_taxonomy: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_usermeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_users: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_wc_admin_notes: Data: 0.06MB + Index: 0.00MB + Engine InnoDB
wp_wc_admin_note_actions: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_category_lookup: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wc_customer_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_download_log: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_order_coupon_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_order_product_lookup: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_wc_order_stats: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_wc_order_tax_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_product_attributes_lookup: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_product_download_directories: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_product_meta_lookup: Data: 0.02MB + Index: 0.09MB + Engine InnoDB
wp_wc_rate_limits: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_reserved_stock: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wc_tax_rate_classes: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_webhooks: Data: 0.02MB + Index: 0.02MB + Engine InnoDB

### Post Type Counts ###

attachment: 1
page: 7
post: 1
product: 1
shop_order: 9
wp_global_styles: 1

### Security ###

Secure connection (HTTPS): ✔
Hide errors from visitors: ✔

### Active Plugins (4) ###

Bookings and Appointments For WooCommerce Premium: by PluginHive – 3.1.1
Pluginhive Multiple Non-Adjacent Bookings Addon: by Pluginhive – 1.2.2
WooCommerce: by Automattic – 7.1.0
WP Crontrol: by John Blackbourn & crontributors – 1.14.0

### Inactive Plugins (5) ###

Advanced Product Fields Extended for WooCommerce: by StudioWombat – 2.2.1
Akismet Anti-Spam: by Automattic – 5.0.1
Breeze: by Cloudways – 2.0.12
Hello Dolly: by Matt Mullenweg – 1.7.2
MalCare WordPress Security Plugin - Malware Scanner, Cleaner, Security Firewall: by MalCare Security – 4.83

### Must Use Plugins (1) ###

Object Cache Pro (MU): by Rhubarb Group – 1.16.3

### Settings ###

API Enabled: –
Force SSL: –
Currency: USD ($)
Currency Position: left
Thousand Separator: ,
Decimal Separator: .
Number of Decimals: 2
Taxonomies: Product Types: external (external)
grouped (grouped)
phive_booking (phive_booking)
simple (simple)
variable (variable)

Taxonomies: Product Visibility: exclude-from-catalog (exclude-from-catalog)
exclude-from-search (exclude-from-search)
featured (featured)
outofstock (outofstock)
rated-1 (rated-1)
rated-2 (rated-2)
rated-3 (rated-3)
rated-4 (rated-4)
rated-5 (rated-5)

Connected to WooCommerce.com: –
Enforce Approved Product Download Directories: ✔

### WC Pages ###

Shop base: #6 - /shop
Cart: #7 - /cart
Checkout: #8 - /checkout
My account: #9 - /my-account
Terms and conditions: ❌ Page not set

### Theme ###

Name: Storefront
Version: 4.2.0
Author URL: https://woocommerce.com/
Child Theme: ❌ – If you are modifying WooCommerce on a parent theme that you did not build personally we recommend using a child theme. See: How to create a child theme
WooCommerce Support: ✔

### Templates ###

Overrides: –

### Admin ###

Enabled Features: activity-panels
analytics
coupons
customer-effort-score-tracks
experimental-products-task
experimental-import-products-task
experimental-fashion-sample-products
shipping-smart-defaults
shipping-setting-tour
homescreen
marketing
multichannel-marketing
mobile-app-banner
navigation
onboarding
onboarding-tasks
remote-inbox-notifications
remote-free-extensions
payment-gateway-suggestions
shipping-label-banner
subscriptions
store-alerts
transient-notices
woo-mobile-welcome
wc-pay-promotion
wc-pay-welcome-page

Disabled Features: minified-js
new-product-management-experience
settings

Daily Cron: ✔ Next scheduled: 2022-11-18 16:39:28 +00:00
Options: ✔
Notes: 58
Onboarding: completed

### Action Scheduler ###

Complete: 92
Oldest: 2022-10-18 00:32:00 +0000
Newest: 2022-11-18 00:12:33 +0000

Pending: 1
Oldest: 2022-11-18 16:16:53 +0000
Newest: 2022-11-18 16:16:53 +0000

### Status report information ###

Generated at: 2022-11-18 00:32:18 +00:00
`

### Isolating the problem

- [X] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [X] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/storefront/).
- [X] I can reproduce this bug consistently using the steps above.

**Comment by senadir** [comment]
I agree, it should also include the `woocommerce_store_api_checkout_order_processed` hook that has a slightly different signature.

I'm struggling to identify the document thou, because the link above no longer work, I will revisit this once I have the correct link.

**Comment by JohnChandlerEdmonton** [comment]
You're right! I don't see anything about this anymore in the woocommerce online documentation.

**Comment by mikejolley** [comment]
Looks like you found a good alternative solution. This doc does not exist any more so Ill close this issue. Thanks.
